### PR TITLE
dotnet/templating#4142 made template configuration model available via API

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
#4142

### Solution
Made template configuration model available via API. Model is read-only. 
To enable creation more refactoring is required.

### Checks:
- [x] Added unit tests
- [ ] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style) - partially

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ConfigModel/ConditionedConfigurationElement.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Evaluates the condition <see cref="Condition"/>.
+        /// </summary>
+        /// <param name="logger">the logger to be used to log the messages during evaluation.</param>
+        /// <param name="variables">the variable collection that will be used to evaluate the condition.</param>
+        /// <returns>true if condition evaluates to true or not specified, otherwise false.</returns>
+        public bool EvaluateCondition(ILogger logger, IVariableCollection variables)
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ConfigModel/ConditionedConfigurationElement.cs:52]: I know this is just code move, but I was wondering if - in spirit of exposing the API - we want to add an overload with `IReadonlyDictionary<string, object> variables` as well

**Review by JanKrivanek — APPROVED** [review]
Generally looks good.

If there is any specific logic that is completely new and needs more detailed review - pleas point that out.

**Comment by vlada-shubina** [comment]
> Generally looks good.
> 
> If there is any specific logic that is completely new and needs more detailed review - pleas point that out.

I would point out couple of things:
- verify exposed APIs: whether naming and nullable annotations make sense. Note about nullable: it would be good to set more restrictive rules, but it needs better validation, so decided to keep it as is and some members are nullable whereas logically they should not. As it is exposed now, it would be bigger impact to change them.
- API change: `IEnumerable<IOperationProvider> ConfigureFromJson(string rawConfiguration, IDirectory templateRoot);` Previously JObject is used which breaks commitment of not exposing JSON classes. Implementations were changed accordingly in a very straightforward way. It is a breaking change however I think the risk of breaking someone here is very low, as having operations components are not likely (if fully supported at all).
- `RunnableProjectConfig` has some refactoring changes

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ConfigModel/ConditionedConfigurationElement.cs (hunk 1 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/ConfigModel/ConditionedConfigurationElement.cs:52]: due to lack of time will keep as is. We can always add this method later on if needed.

**Pull request merged**